readthedocs/notifications/messages.py

```diff
@@ -55,10 +70,20 @@ def get_display_icon_classes(self):
         return " ".join(classes)
 
     def get_rendered_header(self):
-        return self.header.format(**self.format_values)
+        try:
+            return self.header.format(**self.format_values)
+        except KeyError:
+            # There was a key missing
+            log.exception("There was a missing key when formating a header's Message.")
+            return self.header.format_map(defaultdict(str, **self.format_values))
```

Comment:
Since we have control over the notification, I think this should fail hard, our tests should catch these instead of silently failing.

Reply:
We can't catch this by tests since this rendering happens at runtime. This could happen because a `Notification` was created with invalid format values, or because we have change the message over time and started required more format values.

Reply:
The idea here would be to test for each notification we generate on each part of the code base. If we generate a notification when the user does something, we should have a test for it.

Reply:
That's fine, we can test _the current state_, but it could fail still in the future:

1. we have a notification with `body="Go to {url} and put {code}."` and `format_values={"url": "https://..", "code": 1234}`
2. in the future, we change the message to `body="Go to {url} and put {code}. Otherwise, contact support at {email}`

At this point, all the old notifications with only `url` and `code` as format values will start to fail because they are missing the `email` key. This check here is to protect ourselves against this issue.

Reply:
We will find some edge cases as we go that we can improve and write test cases for them and protections like this one against them.